File: src/Pieces/King.java
package Pieces;
import GameElems.Board;

import java.util.List;
import java.util.ArrayList;
public class King extends Piece {
    private boolean hasMoved; // Keep track if the king has moved
    public King(PlayerColor color) {
        super(color, 'K');
        this.hasMoved = false;
    }

    public boolean hasMoved() {
        return hasMoved;
    }



    @Override
    public boolean isValidMove(int startCol, int startRow, int endCol, int endY, Board board) {
        if (wouldThisMovePutKingInCheck(startCol, startRow, endCol, endY, board)) {
            return false;
        }
        // Check if the move is valid for a king (can move one square in any direction)
        int dRows = Math.abs(startCol - endCol);
        int dCol = Math.abs(startRow - endY);

        // The move is valid if the destination square is adjacent and empty or contains an opponent's piece
        boolean isAdjacentMove = dRows <= 1 && dCol <= 1 && (board.getPiece(endCol, endY)== null || board.getPiece(endCol, endY).getColor() != this.getColor());

        // Check for castling
        boolean isKingSideCastle = !hasMoved && dRows == 0 && endY == 6 &&
                board.getPiece(startCol, startRow + 1) == null &&
                board.getPiece(startCol, startRow + 2) == null &&
                !board.isKingInCheck(this.getColor()) &&    // The king cannot be in check while castling
                !wouldThisMovePutKingInCheck(startCol, startRow, startCol, startRow + 1, board); // The king cannot move across check

        if (isKingSideCastle) {
            Piece rook = board.getPiece(startCol, 7);
            isKingSideCastle = rook instanceof Rook && !rook.hasMoved() && rook.getColor()== this.getColor();
        }
        boolean isQueenSideCastling = !hasMoved && dRows == 0 && endY == 2 &&
                board.getPiece(startCol, startRow -1) == null &&
                board.getPiece(startCol, startRow -2) == null &&
                board.getPiece(startCol, startRow -3) == null &&
                !board.isKingInCheck(this.getColor()) &&
                !wouldThisMovePutKingInCheck(startCol, startRow, startCol, startRow - 1, board);  // The king cannot move across check


        if (isQueenSideCastling) {
            Piece rook = board.getPiece(startCol, 0);
            isQueenSideCastling = rook instanceof Rook && !rook.hasMoved() && rook.getColor()== this.getColor();
        }

        return (isAdjacentMove || isKingSideCastle || isQueenSideCastling);
    }

}

File: cpp/Pieces/King.cpp
#include "Piece.h"
#include "GameElems.h"
#include "Rook.h" // Include the Rook header file
#include <vector>

class King : public Piece {
private:
    bool hasMoved;

public:
    King(PlayerColor color) : Piece(color, 'K'), hasMoved(false) {}

    bool hasMoved() override {
        return hasMoved;
    }

    void setHasMoved(bool hasMoved) override {
        this->hasMoved = hasMoved;
    }

    bool isInCheck(int x, int y, Board& board) {
        for (int i = 0; i < 8; ++i) {
            for (int j = 0; j < 8; ++j) {
                Piece* piece = board.getPiece(i, j);
                if (piece != nullptr && piece->getColor() != this->getColor() && piece->isValidMove(i, j, x, y, board)) {
                    return true;
                }
            }
        }
        return false;
    }

    bool isValidMove(int startX, int startY, int endX, int endY, Board& board) override {
        int dx = std::abs(startX - endX);
        int dy = std::abs(startY - endY);

        bool isAdjacentMove = dx <= 1 && dy <= 1 && (board.getPiece(endX, endY) == nullptr || board.getPiece(endX, endY)->getColor() != this->getColor());
        bool isUnderAttack = wouldThisMovePutKingInCheck(startX, startY, endX, endY, board);

        bool isCastlingMove = !hasMoved && dx == 0 && dy == 2 && board.getPiece(startX, startY + dy / 2) == nullptr && board.getPiece(startX, startY + dy) == nullptr;
        if (isCastlingMove) {
            Piece* piece = board.getPiece(startX, startY + dy / 2 * 3);
            isCastlingMove =  && !rook->hasMoved();
        }

        return (isAdjacentMove || isCastlingMove) && !isUnderAttack;
    }

    bool threatensPosition(int x, int y, Board& board) override {
        for (int dx = -1; dx <= 1; ++dx) {
            for (int dy = -1; dy <= 1; ++dy) {
                int newX = x + dx;
                int newY = y + dy;
                if (newX >= 0 && newX < 8 && newY >= 0 && newY < 8 && (board.getPiece(newX, newY) == nullptr || board.getPiece(newX, newY)->getColor() != this->getColor())) {
                    return true;
                }
            }
        }
        return false;
    }

    std::vector<std::vector<int>> generatePossibleMoves(int startX, int startY, Board& board) override {
        std::vector<std::vector<int>> possibleMoves;
        int dx[] = {-1, -1, -1, 0, 0, 1, 1, 1};
        int dy[] = {-1, 0, 1, -1, 1, -1, 0, 1};
        for (int direction = 0; direction < 8; ++direction) {
            int x = startX + dx[direction];
            int y = startY + dy[direction];
            if (x >= 0 && x < 8 && y >= 0 && y < 8 && (board.getPiece(x, y) == nullptr || board.getPiece(x, y)->getColor() != this->getColor())) {
                possibleMoves.push_back({x, y});
            }
        }
        return possibleMoves;
    }
};